### DISCUSSION TITLE
How can I create new users from client?

### DISCUSSION THREAD

**Discussion by ShashankSMayya** [supabase/supabase#20790] [Questions] [answered]
So I have an app where there are admin users who can create the users and then only the users can sign in. 
I have a table called users where there is a field called role and if role is admin then this user can create other admin users and also other roled users. All roles of users are using the same client app(Flutter Web App)

I have admin repo code within Flutter as below
```dart
class AdminRepository {
  final SupabaseClient _client = Supabase.instance.client;

  static const String _users = 'users';

  static const String _teams = 'teams';

  static const String _teamQuery = "*,team:teams(name,id)";

  Future<List<AppUser>> getAllUsers() async {
    return await _client.from(_users).select(_teamQuery).withConverter(
        (data) => data.map((e) => AppUserMapper.fromMap(e)).toList());
  }

  Future<AppUser> addUser(AppUser appUser) async {
    final user = await _client.auth.admin.createUser(AdminUserAttributes(
      email: appUser.email,
      emailConfirm: true,
      userMetadata: appUser.toMap(),
    ));
    return await _client
        .from(_users)
        .select(_teamQuery)
        .eq('id', user.user!.id)
        .single()
        .withConverter((data) => AppUserMapper.fromMap(data));
  }

  Future<AppUser> editUser(AppUser appUser) async {
    return await _client
        .from(_users)
        .update(appUser.toMap())
        .select()
        .single()
        .withConverter((data) => AppUserMapper.fromMap(data));
  }

  Future<void> deleteUser(AppUser appUser) async {
    return await _client.from(_users).delete().eq('id', appUser.id);
  }



  Future<List<Team>> getAllTeams() async {
    return await _client.from(_teams).select().withConverter(
        (data) => data.map((e) => TeamMapper.fromMap(e)).toList());
  }

  Future<void> addTeam(Team team) async {
    return await _client.from(_teams).insert(team.toMap());
  }

  Future<void> editTeam(Team team) async {
    return await _client.from(_teams).update(team.toMap()).eq('id', team.id);
  }

  Future<void> deleteTeam(Team team) async {
    return await _client.from(_teams).delete().eq('id', team.id);
  }
}
```

I also have a trigger in supabase on auth user added  to auth.users table to create a user in public.users table. I wanted to know how can I have admin user create the users from the client side?  When I try to add the user I get 
` User not allowed, statusCode: 401` message.

Can someone help me on how can I achieve this ? I did check on creating a Supabase Client using service_role key but then I am not able to create multiple Supabase client instance in flutter since there is a default supabase instacne with anon key initalized.

**Comment by dshukertjr** [comment, marked as answer]
You should NEVER initialize a Supabase client using the service role key on the client side. Anyone who obtained the service role key can wipe out your database, or read all the data from your database. You should only use it in a server-side environment. 

Because of that reason, you cannot call `supabase.admin.createUser()` from the client side. You can create an [edge functions](https://supabase.com/docs/guides/functions) that calls `supabase.auth.createUser()`, and call the function from the client.

**Reply by ShashankSMayya** [reply]
Thanks a lot for the insights. I'll check that out.
